managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -780,12 +1088,99 @@ private boolean beforeAddEntry(OpAddEntry addOperation) {
         }
     }
 
+    /**
+     * This method should not block the thread, if the buffer is full then use another runnable to fill data
+     * when buffer available.
+     *
+     * @param addOperation
+     */
+    protected synchronized void addToOffload(OpAddEntry addOperation) {
```

Comment:
Why need synchronized here? The OfAddEntry callback always called by single thread?

Reply:
`if (offloadEntryFillTask == null || offloadEntryFillTask.isDone())` may have race condition.

Reply:
If it is just to initialize offloadEntryFillTask, double check lock can be used to avoid locking every time

Reply:
@Renkai Do you mean we have multiple threads to call `addToOffload`? I think only the managed ledger worker thread all this method, otherwise, how to guarantee the add write ordering.